blockchain/src/codec.rs

```diff
@@ -0,0 +1,376 @@
+use crate::shortid::ShortIDVec;
+use crate::utreexo::Proof;
+use crate::{
+    Block, BlockHeader, BlockID, BlockTx, GetBlock, GetInventory, GetMempoolTxs, Inventory,
+    MempoolTxs, Message,
+};
+use curve25519_dalek::ristretto::CompressedRistretto;
+use curve25519_dalek::scalar::Scalar;
+use readerwriter::{Decodable, Encodable, ReadError, Reader, WriteError, Writer};
+use std::convert::TryFrom;
+use zkvm::{merkle, Hash, Signature, Tx};
+
+#[repr(u8)]
+enum MessageType {
+    Block = 0,
+    GetBlock = 1,
+    Inventory = 2,
+    GetInventory = 3,
+    MempoolTxs = 4,
+    GetMempoolTxs = 5,
+}
+
+impl TryFrom<u8> for MessageType {
+    type Error = ReadError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(MessageType::Block),
+            1 => Ok(MessageType::GetBlock),
+            2 => Ok(MessageType::Inventory),
+            3 => Ok(MessageType::GetInventory),
+            4 => Ok(MessageType::MempoolTxs),
+            5 => Ok(MessageType::GetMempoolTxs),
+            _ => Err(ReadError::Custom(
+                format!("unknown message type: {}", value).into(),
+            )),
+        }
+    }
+}
+
+fn read<T>(data: Result<T, ReadError>, label: &'static str) -> Result<T, ReadError> {
```

Comment:
This helper name is super-confusing, as it's mapping the error, not reading anything.
Do we really need it? It maps ReadError into ReadError, maybe leave it as is? 

If you really want to re-wrap it, i'd make a func `rewrap_err(ReadError,..)->ReadError` and use it as `foo().map_err(|e| rewrap_err(e, b"foo failed"))?`

Reply:
Agree with you, this function is excess